File: check_run/check_run/report/positive_pay/positive_pay.js
// Copyright (c) 2022, AgriTheory and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Positive Pay"] = {
	"filters": [
		{
			fieldname: "bank_account",
			label: __("Bank Account"),
			fieldtype: "Link",
			options: 'Bank Account',
			reqd: 1,
		},
		{
			fieldname: "start_date",
			label: __("Start Date"),
			fieldtype: "Date",
			default: moment().date(0).startOf('month').format(),
			reqd: 1,
		},
		{
			fieldname: "end_date",
			label: __("End Date"),
			fieldtype: "Date",
			default: moment().date(0).format(),
			reqd: 1,
		},
	]
}

File: check_run/check_run/report/positive_pay/positive_pay.py
# Copyright (c) 2022, AgriTheory and contributors
# For license information, please see license.txt

import frappe


def execute(filters=None):
	return get_columns(filters), get_data(filters)


def get_columns(filters):
	return [
		{
			"label": frappe._("Check Date"),
			"fieldname": "check_date",
			"fieldtype": "Date",
			"width": "150px",
		},
		{
			"label": frappe._("Check Number"),
			"fieldname": "check_number",
			"fieldtype": "Data",
			"width": "200px",
		},
		{
			"label": frappe._("Party Name"),
			"fieldname": "party_name",
			"fieldtype": "Data",
			"width": "400px",
		},
	]


def get_data(filters):
	pe = frappe.qb.DocType("Payment Entry")
	mop = frappe.qb.DocType("Mode of Payment")
	return (
		frappe.qb.from_(pe)
		.inner_join(mop)
		.on(pe.mode_of_payment == mop.name)
		.select(
			(pe.reference_no).as_("check_number"),
			(pe.reference_date).as_("check_date"),
			pe.party_name,
		)
		.where(pe.reference_date >= filters.start_date)
		.where(pe.reference_date <= filters.end_date)
		.where(pe.bank_account >= filters.bank_account)
		.where(mop.type == "Bank")
		.where(pe.docstatus == 1)
		.orderby(pe.reference_date)
		.run(as_dict=True)
	)
